src/agents/graph.py

```diff
@@ -0,0 +1,342 @@
+                    [
+                        AIMessage(
+                            content=f"Task decomposed into subtasks and assigned to agents: {state.subtasks}",
+                            name=PLANNER,
+                        )
+                    ]
+                ),
+                "error": None,
+            }
+        except Exception as e:
+            logger.error(f"Error in planning: {e}")
+            return {
+                "error": str(e),
+                "messages": [
+                    AIMessage(
+                        content=f"Error in planning: {e}",
+                        name="Planner",
+                    )
+                ],
+            }
+
+    def common_node(self, state: AgentState) -> dict[str, Any]:
+        """Breaks down the given user query into sub-tasks."""
+
+        prompt = ChatPromptTemplate.from_messages(
+            [
+                (
+                    "system",
+                    "Given the conversation and the last user query, "
+                    "you are tasked with generating a response. ",
+                ),
+                MessagesPlaceholder(variable_name="messages"),
+                ("human", "query: {query}"),
+            ]
+        )
+        chain = prompt | self.model.llm
+
+        for subtask in state.subtasks:
+            if subtask.assigned_to == COMMON and subtask.status != "completed":
+                try:
+                    response = chain.invoke(
+                        {
+                            "messages": state.messages,
+                            "query": subtask.description,
+                        },
+                        config={"callbacks": [handler]},
+                    ).content
+                    subtask.complete()
+                    return {
+                        "messages": [
+                            AIMessage(
+                                content=response,
+                                name=COMMON,
+                            )
+                        ],
+                    }
+                except Exception as e:
+                    logger.error(f"Error in common node: {e}")
+                    return {
+                        "error": str(e),
+                        "messages": [
+                            AIMessage(
+                                content=f"Error occurred: {e}",
+                                name=COMMON,
+                            )
+                        ],
+                    }
+        return {
+            "messages": [
+                AIMessage(
+                    content="All my subtasks are already completed.",
+                    name=COMMON,
+                )
+            ]
+        }
+
+    def generate_final_response(self, state: AgentState) -> dict[str, Any]:
+        """Generate the final response."""
+
+        prompt = ChatPromptTemplate.from_messages(
+            [
+                MessagesPlaceholder(variable_name="messages"),
```

Comment:
Supervisor agent should have an ability to fail properly in case other agents can't provide relevant information. Therefore it should probably analyze provided information and make a decision whether this information is sufficient and it can generate a comprehensive response or it should fail. I suspect, additional prompting should be done in the finalizer. Here is an example where an agent fails in doing that:
[request_response1.txt](https://github.com/user-attachments/files/16798505/request_response1.txt)